Ardakilic keymap for crkbd added

Ardakilic opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

My keymap for Corne keyboard. This keymap is optimized for typing Turkish while covering my needs for programming, and is similar to #7643 and #7151

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

- None

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/crkbd/keymaps/ardakilic/config.h:
```diff
@@ -0,0 +1,55 @@
+
+#pragma once
+
+//#define USE_MATRIX_I2C
```
depricated.
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/config.h:
```diff
@@ -0,0 +1,55 @@
+// #define MASTER_RIGHT
+// #define EE_HANDS
+
+#define USE_SERIAL_PD2
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/config.h:
```diff
@@ -0,0 +1,55 @@
+#define OLED_FONT_H "keyboards/crkbd/lib/glcdfont.c"
+
+
+// Size optimisations
+#ifndef NO_DEBUG
+#define NO_DEBUG
+#endif // !NO_DEBUG
+#if !defined(NO_PRINT) && !defined(CONSOLE_ENABLE)
+#define NO_PRINT
+#endif // !NO_PRINT
```
If console isn't enabled, these are automatically defined, no need for them. 
```suggestion
// Size optimisations
```

Ardakilic replied:
@drashna I have added CONSOLE_ENABLE = no in rules.mk and deleted these, but it now says

```
 * The firmware is too large! 28812/28672 (140 bytes over)
 [ERRORS]

make[1]: *** [check-size] Error 1
make: *** [crkbd:ardakilic:avrdude] Error 1
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/config.h:
```diff
@@ -0,0 +1,55 @@
+#define NO_PRINT
+#endif // !NO_PRINT
+
+#define NO_ACTION_MACRO
+#define NO_ACTION_ONESHOT
+#define NO_ACTION_FUNCTION
```
macro and function disableds are defined at keyboard level.
```suggestion
#define NO_ACTION_ONESHOT
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,482 @@
+
+// Layers
+enum {
+  _QWERTY = 0,
+  _LOWER = 1,
+  _RAISE = 2,
+  _SPACE = 3,
+  _NUMPAD = 4,
+  _SODA = 5,
+  _ADJUST = 6
```
The numbers are not needed.  Enums start at 0, and automatically increment. 
```suggestion
  _QWERTY,
  _LOWER,
  _RAISE2,
  _SPACE,
  _NUMPAD,
  _SODA,
  _ADJUST,
```

Ardakilic replied:
@drashna any reason you've renamed _RAISE to _RAISE2, or was that just a typo ? 

I had problems while compiling due to this and had to specifically set `0` at least to `_QWERTY `back then (using macOS if that would matter), then I thought I'd just put into them all.

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,482 @@
+
+
+#ifdef OLED_DRIVER_ENABLE
+#include <stdio.h>
```
```suggestion
```

Just a heads up, snprintf/sprintf uses up 1.5kb of firmware space, compared to something like this: 
```c
    uint8_t n = get_current_wpm();
    char wpm_counter[4];
    wpm_counter[3] = '\0';
    wpm_counter[2] = '0' + n % 10;
    wpm_counter[1] = ( n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
    wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
    oled_write_P(PSTR("WPM: "), false);
    oled_write(wpm_counter, false);
```

Ardakilic replied:
Thanks @drashna , this is awesome!

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,482 @@
+
+void oled_render_layer_state(void) {
+    oled_write_P(PSTR("Layer: "), false);
+    switch (biton32(layer_state)) {
```
```suggestion
    switch (get_higest_layer(layer_state)) {
```

fauxpark commented on keyboards/crkbd/keymaps/ardakilic/README.md:
```diff
@@ -0,0 +1,8 @@
+Ardakilic's Corne Layout
+--------
```
```suggestion
# Ardakilic's Corne Layout
```
Please also rename this file to lowercase.

fauxpark commented on keyboards/crkbd/keymaps/ardakilic/README.md:
```diff
@@ -0,0 +1,7 @@
+
+This layout is optimised to use Turkish characters on Corne on macOS.
+
+Also, some size optimisations enabled in ´rules.mk´ and ´config.h´
```
```suggestion
Also, some size optimisations enabled in `rules.mk` and `config.h`.
```

Ardakilic commented:
`CONSOLE_ENABLE=no` did not help for my case. My initial lines for size optimisation work though.

I have also added

```c
EXTRAFLAGS+=-flto
```

to my `rules.mk` file, and added

```c
#define NO_ACTION_MACRO
#define NO_ACTION_FUNCTION
```

to my config.h file. Now it fits. 

Please check my last commits. I wonder why disabling console does not reduce the size like my initial commit though.

drashna commented:
No need to add those.  `LTO_ENABLE = yes` does both of those, actually

drashna commented on keyboards/crkbd/keymaps/ardakilic/rules.mk:
```diff
@@ -0,0 +1,13 @@
+# Size optimisations
+CONSOLE_ENABLE = no
+COMMAND_ENABLE = no
+EXTRAFLAGS+=-flto
```
```suggestion
LTO_ENABLE = no
SPACE_CADET_ENABLE = no
GRAVE_ESCAPE_ENABLE = no
```

drashna commented:
Also, you  may want to remove snprint/sprint, because this can take up a LOT of space in the firmwar (1.5kb or more). 

For rendering numbers, this may be more verbose, but ...

```c
    uint8_t n = get_current_wpm();
    char wpm_counter[4];
    wpm_counter[3] = '\0';
    wpm_counter[2] = '0' + n % 10;
    wpm_counter[1] = (n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
    wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
    oled_write_P(PSTR("WPM: "), false);
    oled_write(wpm_counter, false);
```

This renders a 3 digit character, using the WPM value, and removes empty values (eg `  5` vs `005`).  And this uses ~1.5KB less than using snprintf to do the same.

Ardakilic commented:
OK, anythingh else I need to address to? cc/ @drashna @fauxpark

Ardakilic commented:
Are you waiting any further changes from me ?

Ardakilic commented:
Apologies for the bump @drashna , this PR had been stale while awaiting for review, and bot auto closed it. @fauxpark approved this, any reason this is left dormant, or won't this be merged ?

Thanks,

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,494 @@
+};
+
+
+#ifdef OLED_DRIVER_ENABLE
```
```suggestion
#ifdef OLED_ENABLE
```

Ardakilic replied:
done in https://github.com/qmk/qmk_firmware/pull/13745/commits/cedae37b86761fbd6bbf7d4a689181e015e5f946

krikun98 replied:
Hey, this didn't appear to get merged to the main branch.

Ardakilic replied:
Aha, nice catch @krikun98 , thanks! I've just created #16284 regarding this.

krikun98 replied:
@Ardakilic, would you happen to have a branch/hex with the working animation? Even after fixing this bug the WPM counter is just stuck at 0 on my board.

Ardakilic replied:
Hey @krikun98 , sadly I sold my current Corne (classic) to one of my friends, but a low profile PCB is produced from factory and is on the way. I can't sadly test right now.

However, I took the animation from here: https://github.com/nwii/oledbongocat , it should work nicely.

Also, 

```c
char wpm_str[10];
void oled_task_user(void) {
sprintf(wpm_str, "WPM:%03d", get_current_wpm())
oled_write(wpm_str, false);
}
```

lines should suffice for wpm counter, regardless of the animation.

Also check this, it provides two types of examples for WPM: https://docs.qmk.fm/#/squeezing_avr?id=oled-tweaks

krikun98 replied:
The issue was the lack of `#define SPLIT_WPM_ENABLE` in the code, the central half has no way to communicate WPM to the peripheral otherwise. This should probably be included?

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,494 @@
+//    oled_write_P(crkbd_logo, false);
+// }
+
+void oled_task_user(void) {
```
```suggestion
bool oled_task_user(void) {
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,494 @@
+  }
+
+  // update keylog
+  snprintf(keylog_str, sizeof(keylog_str), "%dx%d, k%2d : %c",
+           record->event.key.row, record->event.key.col,
+           keycode, name);
```
https://docs.qmk.fm/#/squeezing_avr?id=oled-tweaks

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,494 @@
+
+        //oled_render_logo();
+    }
+}
```
```suggestion
    return false;
}
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/keymap.c:
```diff
@@ -0,0 +1,494 @@
+        // oled_write(wpm_str, false);                       // writes wpm on top left corner of string
+
+        // This way it's more size efficient than calling sprintf
+        uint8_t n = get_current_wpm();
+        char wpm_counter[4];
+        wpm_counter[3] = '\0';
+        wpm_counter[2] = '0' + n % 10;
+        wpm_counter[1] = (n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
+        wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
+        oled_write_P(PSTR("WPM: "), false);
+        oled_write(wpm_counter, false);
```
```suggestion
        oled_write_P(PSTR("WPM: "), false);
        oled_write(get_u8_str(get_current_wpm(), ' '), false);
```

drashna commented on keyboards/crkbd/keymaps/ardakilic/rules.mk:
```diff
@@ -0,0 +1,15 @@
+OLED_DRIVER_ENABLE  = yes
```
```suggestion
OLED_ENABLE  = yes
```

Ardakilic commented:
Thanks for the review and merge! 🙏
